File: ios/Classes/classifiers/CardioBridge.swift
import Flutter
import UIKit

// MARK: - CardioBridge

class CardioBridge: NSObject {

    // MARK: Static state (used by C callbacks which cannot capture `self`)

    /// The currently active bridge. C callbacks reach the instance through this
    /// static weak reference because `clCCardio_Set*Event` callbacks have no
    /// `void* context` parameter.
    private static weak var activeBridge: CardioBridge?

    // MARK: Stream handlers

    let cardioDataHandler = DeviceStreamHandler(channelId: "neiry_kit/events/cardioData")
    let cardioPpgHandler = DeviceStreamHandler(channelId: "neiry_kit/events/ppgData")
    let cardioCalibratedHandler = DeviceStreamHandler(channelId: "neiry_kit/events/cardioCalibratedEvent")

    /// Returns all (channelId, handler) pairs for EventChannel registration in the plugin.
    func allStreamHandlers() -> [(String, FlutterStreamHandler)] {
        return [
            (cardioDataHandler.channelId, cardioDataHandler),
            (cardioPpgHandler.channelId, cardioPpgHandler),
            (cardioCalibratedHandler.channelId, cardioCalibratedHandler),
        ]
    }

    // MARK: Instance state

    private var cardio: OpaquePointer?

    // MARK: - Public API

    /// Creates an uncalibrated Cardio classifier for the given device handle.
    func create(device: OpaquePointer) throws {
        if cardio != nil { unregisterCallbacks() }
        var error = clCError()
        cardio = clCCardio_Create(device, &error)
        try checkCError(error)
        try registerCallbacks()
    }

    /// Creates a calibrated Cardio classifier for the given device handle.
    /// If `calibrationData` is provided, it is imported into the calibrator before creating.
    func createCalibrated(device: OpaquePointer, calibrationData: [String: Any]?) throws {
        if cardio != nil { unregisterCallbacks() }
        guard let calibrator = clCNFBCalibrator_CreateOrGet(device) else {
            throw FlutterError(code: "NULL_HANDLE", message: "clCNFBCalibrator_CreateOrGet returned nil", details: nil)
        }
        if let calibrationData = calibrationData {
            var data = clCIndividualNFBData()
            if let ts = calibrationData["ts"] as? Int {
                data.timestampMilli = Int64(ts)
            }
            if let failReasonRaw = calibrationData["failReason"] as? Int {
                data.failReason = clCIndividualNFBCalibrationFailReason(rawValue: UInt32(failReasonRaw))
            }
            if let v = calibrationData["individualFrequency"] as? Double { data.individualFrequency = Float(v) }
            if let v = calibrationData["individualPeakFrequency"] as? Double { data.individualPeakFrequency = Float(v) }
            if let v = calibrationData["individualPeakFrequencyPower"] as? Double { data.individualPeakFrequencyPower = Float(v) }
            if let v = calibrationData["individualPeakFrequencySuppression"] as? Double { data.individualPeakFrequencySuppression = Float(v) }
            if let v = calibrationData["individualBandwidth"] as? Double { data.individualBandwidth = Float(v) }
            if let v = calibrationData["individualNormalizedPower"] as? Double { data.individualNormalizedPower = Float(v) }
            if let v = calibrationData["lowerFrequency"] as? Double { data.lowerFrequency = Float(v) }
            if let v = calibrationData["upperFrequency"] as? Double { data.upperFrequency = Float(v) }
            var importError = clCError()
            clCNFBCalibrator_ImportIndividualNFBData(calibrator, &data, &importError)
            try checkCError(importError)
        }
        var error = clCError()
        cardio = clCCardio_CreateCalibrated(device, calibrator, &error)
        try checkCError(error)
        try registerCallbacks()
    }

    /// Unregisters all callbacks and releases the Cardio handle.
    /// There is no `clCCardio_Destroy` in the SDK — disposal is callback-only.
    func dispose() {
        unregisterCallbacks()
        cardio = nil
    }

    // MARK: - Callbacks

    private func registerCallbacks() throws {
        guard let cardio = cardio else { return }
        CardioBridge.activeBridge = self

        var e1 = clCError()
        clCCardio_SetOnIndexesUpdateEvent(cardio, { _, data in
            guard let bridge = CardioBridge.activeBridge,
                  let data = data else { return }
            let d = data.pointee
            let map: [String: Any] = [
                "ts":               d.timestampMilli,
                "heartRate":        d.heartRate,
                "stressIndex":      d.stressIndex,
                "kaplanIndex":      d.kaplanIndex,
                "hasArtifacts":     d.hasArtifacts,
                "skinContact":      d.skinContact,
                "motionArtifacts":  d.motionArtifacts,
                "metricsAvailable": d.metricsAvailable,
            ]
            bridge.cardioDataHandler.send(map)
        }, &e1)
        try checkCError(e1)

        var e2 = clCError()
        clCCardio_SetOnPPGDataEvent(cardio, { _, ppgData in
            guard let bridge = CardioBridge.activeBridge,
                  let ppgData = ppgData else { return }
            let count = clCPPGTimedData_GetCount(ppgData)
            var values: [Float] = []
            var timestamps: [UInt64] = []
            for i in 0..<count {
                values.append(clCPPGTimedData_GetValue(ppgData, i))
                timestamps.append(clCPPGTimedData_GetTimestampMilli(ppgData, i))
            }
            let map: [String: Any] = [
                "sampleCount": count,
                "values":      values,
                "timestamps":  timestamps,
            ]
            bridge.cardioPpgHandler.send(map)
        }, &e2)
        try checkCError(e2)

        var e3 = clCError()
        clCCardio_SetOnCalibratedEvent(cardio, { _ in
            guard let bridge = CardioBridge.activeBridge else { return }
            bridge.cardioCalibratedHandler.send([:])
        }, &e3)
        try checkCError(e3)
    }

    private func unregisterCallbacks() {
        guard let cardio = cardio else { return }
        var e = clCError()
        clCCardio_SetOnIndexesUpdateEvent(cardio, nil, &e)
        clCCardio_SetOnPPGDataEvent(cardio, nil, &e)
        clCCardio_SetOnCalibratedEvent(cardio, nil, &e)
        if CardioBridge.activeBridge === self {
            CardioBridge.activeBridge = nil
        }
    }
}

File: android/src/main/kotlin/com/neiry/neiry_kit/CardioBridge.kt
package com.neiry.neiry_kit

import io.flutter.plugin.common.EventChannel

/**
 * Kotlin bridge for the `clCCardio` classifier. Owns the native Cardio handle
 * and exposes factory methods (plain and calibrated) plus stream handlers for
 * the three event channels: cardioState, cardioPpg, and cardioCalibratedEvent.
 *
 * Mirrors [CardioBridge.swift] in behavior.
 */
class CardioBridge(private val nativeBridge: NativeBridge) {

    private var handle: Long = 0L

    // ── Stream handlers ───────────────────────────────────────────────────────

    inner class CardioStreamHandler(
        private val setSink: (EventChannel.EventSink?) -> Unit
    ) : EventChannel.StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            setSink(events)
        }
        override fun onCancel(arguments: Any?) {
            setSink(null)
        }
    }

    val cardioStateHandler      = CardioStreamHandler { nativeBridge.nativeSetCardioStateSink(it) }
    val cardioPpgHandler        = CardioStreamHandler { nativeBridge.nativeSetCardioPpgSink(it) }
    val cardioCalibratedHandler = CardioStreamHandler { nativeBridge.nativeSetCardioCalibratedSink(it) }

    /** Returns all (channelId, handler) pairs for EventChannel registration in the plugin. */
    fun allStreamHandlers(): List<Pair<String, EventChannel.StreamHandler>> = listOf(
        "neiry_kit/events/cardioData"           to cardioStateHandler,
        "neiry_kit/events/ppgData"              to cardioPpgHandler,
        "neiry_kit/events/cardioCalibratedEvent" to cardioCalibratedHandler,
    )

    // ── Factory methods ───────────────────────────────────────────────────────

    /** Creates an uncalibrated Cardio classifier for the given device handle. */
    fun create(deviceHandle: Long) {
        if (handle != 0L) dispose()
        try {
            handle = nativeBridge.nativeCreateCardio(deviceHandle)
        } catch (e: RuntimeException) {
            throw parseSdkError(e.message ?: "255|Unknown error")
        }
    }

    /**
     * Creates a calibrated Cardio classifier for the given device handle.
     * If [calibrationData] is non-null, its fields are imported into the
     * calibrator before creating the classifier.
     */
    fun createCalibrated(deviceHandle: Long, calibrationData: Map<String, Any>?) {
        if (handle != 0L) dispose()
        val hasData = calibrationData != null
        val ts                                 = (calibrationData?.get("ts") as? Number)?.toLong()    ?: 0L
        val failReason                         = (calibrationData?.get("failReason") as? Number)?.toInt()  ?: 0
        val individualFrequency                = (calibrationData?.get("individualFrequency") as? Number)?.toFloat()             ?: 0f
        val individualPeakFrequency            = (calibrationData?.get("individualPeakFrequency") as? Number)?.toFloat()         ?: 0f
        val individualPeakFrequencyPower       = (calibrationData?.get("individualPeakFrequencyPower") as? Number)?.toFloat()    ?: 0f
        val individualPeakFrequencySuppression = (calibrationData?.get("individualPeakFrequencySuppression") as? Number)?.toFloat() ?: 0f
        val individualBandwidth                = (calibrationData?.get("individualBandwidth") as? Number)?.toFloat()             ?: 0f
        val individualNormalizedPower          = (calibrationData?.get("individualNormalizedPower") as? Number)?.toFloat()       ?: 0f
        val lowerFrequency                     = (calibrationData?.get("lowerFrequency") as? Number)?.toFloat()                  ?: 0f
        val upperFrequency                     = (calibrationData?.get("upperFrequency") as? Number)?.toFloat()                  ?: 0f

        try {
            handle = nativeBridge.nativeCreateCardioCalibrated(
                deviceHandle,
                hasData,
                ts,
                failReason,
                individualFrequency,
                individualPeakFrequency,
                individualPeakFrequencyPower,
                individualPeakFrequencySuppression,
                individualBandwidth,
                individualNormalizedPower,
                lowerFrequency,
                upperFrequency,
            )
        } catch (e: RuntimeException) {
            throw parseSdkError(e.message ?: "255|Unknown error")
        }
    }

    /** Unregisters callbacks and clears the Cardio handle. */
    fun dispose() {
        if (handle != 0L) {
            nativeBridge.nativeDisposeCardio(handle)
            handle = 0L
        }
    }
}
